node/src/Unison/Runtime/ResourcePool.hs

```diff
@@ -0,0 +1,82 @@
+module Unison.Runtime.ResourcePool where
+
+import Data.Time (UTCTime, getCurrentTime, addUTCTime)
+import qualified Control.Concurrent as CC
+import qualified Control.Concurrent.Map as M
+import qualified Control.Concurrent.STM.TQueue as TQ
+import qualified Data.Hashable as H
+import qualified Control.Monad.STM as STM
+import qualified Data.IORef as IORef
+
+-- acquire returns the resource, and the cleanup action ("finalizer") for that resource
+data Pool p r = Pool { acquire :: p -> Int -> IO (r, IO ()) }
+
+type ResourceKey p = (p,CC.ThreadId)
+type Cache p r = M.Map (ResourceKey p) (r, UTCTime, IO ())
+type ReaperQueue p = TQ.TQueue (ResourceKey p)
+
+incRef :: Num t => t -> (t, ())
+incRef a = (a+1,())
+decRef :: Num t => t -> (t, ())
+decRef a = (a-1,())
+
+addResourceToMap :: (Ord p, H.Hashable p) =>  (r -> IO ()) -> Cache p r -> p -> r -> Int -> Int -> IO CC.ThreadId -> IORef.IORef Int -> ReaperQueue p -> IO ()
+addResourceToMap release pool p r wait maxPoolSize getThread poolSizeR queue = do
+  now <- getCurrentTime
+  poolSize <- IORef.readIORef poolSizeR
+  threadId <- getThread
+  let expiry = addUTCTime (fromIntegral wait) now
+  let key = (p,threadId)
+  if wait > 0 && (poolSize < maxPoolSize) then
+    -- add the resource to the pool with the release
+    M.insert key (r, expiry, (release r)) pool
+        >> IORef.atomicModifyIORef poolSizeR incRef
+        >> STM.atomically (TQ.writeTQueue queue key)
+    else release r
+
+_acquire :: (Ord p, H.Hashable p) => (p -> IO r) -> (r -> IO ()) -> Cache p r -> Int -> IORef.IORef Int -> ReaperQueue p -> IO CC.ThreadId -> p -> Int -> IO (r, IO ())
+_acquire acquire release pool maxPoolSize poolSizeR queue getThread p wait = do
+  threadId <- getThread
+  m <- M.lookup (p,threadId) pool
+  r <- case m of
+        Just (r, _, _) -> M.delete (p,threadId) pool
+          >> IORef.atomicModifyIORef poolSizeR decRef
+          >> return r
+        Nothing -> acquire p
+  return (r, (addResourceToMap release pool p r wait maxPoolSize getThread poolSizeR queue))
+
+cleanPool :: (Ord p, H.Hashable p) => Cache p r -> ReaperQueue p -> IO ()
+cleanPool pool queue = do
```

Comment:
Okay, so I realized that the current polling logic is not correct either. The issue is that if we are allowing separate TTLs for each resource, we can't be guaranteed that the thing at the front of the queue is the msg closest to expiration. Thus, using that as the stopping condition for the drain loop is not correct. There might be an expired resource several messages behind the one you gave up on. Tell me if I am missing anything with my reasoning here.

One way to fix this would be to drain items from the queue into a heap, ordered by expiration time, but this is overkill.

The other thing is I don't think the reaper thread will ever be GC'd, even when the pool is no longer being referenced by anyone else. The issue is that it *always* wakes up every second. That is kind of bad. It would be better to kill that thread when the pool (and hence the `TQueue`) is no longer being referenced by anyone else. I have [a question out on that][so].

[so]: http://stackoverflow.com/questions/36310299/will-a-haskell-thread-be-gcd-if-blocked-indefinitely-on-an-mvar-chan-tqueue-rea

I think having separate TTL for each resource is overkill. Just have a single TTL for all resources in the `Pool`. If you really need multiple settings, just use different pools. For this use case, I think [the logic I gave above works](https://github.com/unisonweb/unison/pull/38/files#r57505418), though I want to get a definite answer on [whether the reaper thread will get GC'd if it is blocked indefinitely][so].